qiita_db/search.py

```diff
@@ -194,7 +200,8 @@ def __call__(self, searchstr, user):
         Metadata column names and string searches are case-sensitive
         """
         study_sql, sample_sql, meta_headers = \
-            self._parse_study_search_string(searchstr, True)
```

Comment:
Decomposition doesn't necessarily mean doing this in python, and python is not necessarily slower. What I'm recommending is to break the function into smaller blocks as it will be much easier to verify that the individual pieces function as expected. It is very difficult to verify that a large function with complex branching logic is correct.